README.md
# CE2727A simple driver

[![Component Registry](https://components.espressif.com/components/latonita/ce2727a-driver/badge.svg)](https://components.espressif.com/components/latonita/ce2727a-driver) [![Example build](https://github.com/latonita/ce2727a-driver/actions/workflows/build_example.yml/badge.svg)](https://github.com/latonita/ce2727a-driver/actions/workflows/build_example.yml)

This repository contains an ESP-IDF driver for a CE2727A power meter connected over UART (RS485).

## Using the component

Run the following command in your ESP-IDF project to install this component:
```bash
idf.py add-dependency "latonita/ce2727a-driver"
```

## Example

To run the provided example, create it as follows:

```bash
idf.py create-project-from-example "latonita/ce2727a-driver:ce2727a-example"
```

Then build as usual:
```bash
cd ce2727a-example
idf.py build
```

And flash it to the board:
```bash
idf.py -p PORT flash monitor
```

....todo....

## License

This component is provided under Apache 2.0 license, see [LICENSE](LICENSE.md) file for details.

## Contributing

Please check [CONTRIBUTING.md](CONTRIBUTING.md) for contribution guidelines.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Command: Request general information
- Command: Request date/time

## [0.2.0] - 2023-09-15

- Initial release


frames.h
#pragma once

#include <stdio.h>

#pragma pack(1)
typedef struct
{
    uint8_t two;
    uint8_t length;
    uint32_t address;
    uint32_t password;
    uint8_t com;
    uint8_t id;
} ce2727a_frame_header_t;

typedef struct
{
    ce2727a_frame_header_t header;
    uint16_t crc16;
} ce2727a_request_command_t;

typedef struct
{
    ce2727a_frame_header_t header;
    uint8_t second;
    uint8_t minute;
    uint8_t hour;
    uint8_t day;
    uint8_t month;
    uint8_t year;
    uint8_t weekdayAndSummer;
    uint8_t dstAllowed;
    uint8_t correctionSecondsLeft;
    uint16_t crc16;
} ce2727a_response_date_time_t;

typedef struct
{
    ce2727a_frame_header_t header;
    uint32_t activePower;
    uint16_t crc16;
} ce2727a_response_active_power_t;

typedef struct
{
    ce2727a_frame_header_t header;
    uint8_t currentTariff;
    uint32_t total;
    uint32_t t1;
    uint32_t t2;
    uint32_t t3;
    uint32_t t4;
    uint16_t crc16;
} ce2727a_response_consumed_energy_t;
#pragma pack(0)


include/ce2727a.h
#pragma once

#include "esp_err.h"
#include "driver/uart.h"

#ifdef __cplusplus
extern "C"
{
#endif

    /**
     * @brief CE2727A driver configuration
     */
    typedef struct
    {
        uart_config_t *uart_config_override; /*!< UART port configuration override parameters. defaults: 9600,8E1 */
        uint8_t uart_port;                   /*!< UART port number*/
        uint8_t uart_tx_pin;                 /*!< UART TX pin */
        uint8_t uart_rx_pin;                 /*!< UART RX pin */
        uint8_t uart_de_pin;                 /*!< UART RTS pin for RS485 DE=!RE */
        uint32_t addr;                       /*!< Target power meter address, default is 0x0 - broadcasting to all meters  */
        uint32_t password;                   /*!< Target power meter password, default is 0x0 */
        uint8_t *rx_buffer;                  /*!< Receive buffer. Shall be enough to hold required responses */
        uint16_t rx_buffer_size;             /*!< Size of receive buffer */
    } ce2727a_config_t;

    /**
     * @brief CE2727A readings
     */
    typedef struct
    {
        uint8_t active_tariff; /*!< Current active tariff [1,2,3,4] */
        uint32_t total;        /*!< Cumulative energy consumed on all tariffs */
        uint32_t t1;           /*!< Cumulative energy for tariff 1 */
        uint32_t t2;           /*!< Cumulative energy for tariff 2 */
        uint32_t t3;           /*!< Cumulative energy for tariff 3 */
        uint32_t t4;           /*!< Cumulative energy for tariff 4 */
    } ce2727a_readings_energy_t;

    /**
     * @brief CE2727A driver handle
     */
    typedef struct ce2727a_driver_state_t *ce2727a_handle_t;

    /**
     * @brief Initialize the CE2727A driver
     *
     * @param[in] config Pointer to the configuration struct. The driver makes a copy, so can point to a local variable.
     * @param[out] out_handle  Pointer to a variable to receive the driver handle.
     * @return esp_err_t  ESP_OK on success, ESP_ERR_NO_MEM if out of memory.
     */
    esp_err_t ce2727a_init(const ce2727a_config_t *config, ce2727a_handle_t *out_handle);

    /**
     * @brief Deinitialize the CE2727A driver
     *
     * @param[in] handle Driver handle obtained from ce2727a_init(), or NULL
     * @return esp_err_t  ESP_OK on success.
     */
    esp_err_t ce2727a_deinit(ce2727a_handle_t handle);

    /**
     * @brief Get active power
     *
     * Get current active power
     *
     * @param[in] handle Driver handle obtained from ce2727a_init()
     * @param[out] active_power Active power
     * @param[in] timeout_ms Timeout in milliseconds for receiving a response from the meter
     * @return esp_err_t
            - ESP_OK on success
            - ESP_ERR_TIMEOUT if no message was received within the timeout
            - ESP_ERR_INVALID_ARG if the buffer too short
            - ESP_ERR_INVALID_RESPONSE
            - ESP_ERR_INVALID_CRC
            - ESP_ERR_INVALID_STATE Issues sending data
     */
    esp_err_t ce2727a_get_active_power(ce2727a_handle_t handle, uint32_t *active_power, uint16_t timeout_ms);

    /**
     * @brief Get energy readings
     *
     * Get energy readings
     *
     * @param[in] handle Driver handle obtained from ce2727a_init()
     * @param[out] readings Destination for readings
     * @param[in] timeout_ms Timeout in milliseconds for receiving a response from the meter
     * @return esp_err_t
            - ESP_OK on success
            - ESP_ERR_TIMEOUT if no message was received within the timeout
            - ESP_ERR_INVALID_ARG if the buffer too short
            - ESP_ERR_INVALID_RESPONSE
            - ESP_ERR_INVALID_CRC
            - ESP_ERR_INVALID_STATE Issues sending data
     */
    esp_err_t ce2727a_get_energy(ce2727a_handle_t handle, ce2727a_readings_energy_t *readings, uint16_t timeout_ms);

#ifdef __cplusplus
}
#endif


examples/ce2727a-example/main/ce2727a-example.c
#include <stdio.h>
#include <freertos/FreeRTOS.h>
// #include <freertos/task.h>
#include "esp_log.h"
#include "ce2727a.h"
#include "driver/uart.h"

static const char *TAG = "ce2727a-example";

#define BUF_SIZE 128
uint8_t rx_buffer[BUF_SIZE];

void app_main(void)
{
    ESP_LOGI(TAG, "CE2727a example started");

    ce2727a_config_t config = {
        .uart_config_override = NULL,
        .uart_port = UART_NUM_2,
        .uart_rx_pin = 16,
        .uart_tx_pin = 17,
        .uart_de_pin = 18,
        .addr = 0x0,
        .password = 0x0,
        .rx_buffer = rx_buffer,
        .rx_buffer_size = BUF_SIZE};

    ESP_LOGI(TAG, "Init ce2727a driver");
    ce2727a_handle_t handle = NULL;
    ESP_ERROR_CHECK(ce2727a_init(&config, &handle));

    while (true)
    {
        ce2727a_readings_energy_t readings = {0};
        uint32_t active_power = 0;

        esp_err_t err = ESP_OK;

        ESP_LOGI(TAG, "Requesting active power...");
        err = ce2727a_get_active_power(handle, &active_power, 1000);
        if (err == ESP_OK)
        {
            ESP_LOGI(TAG, "Got active power: %lu Watts", active_power);
        }
        else
        {
            ESP_LOGE(TAG, "Error getting active power: 0x%x", err);
        }

        ESP_LOGI(TAG, "Requesting readings...");
        err = ce2727a_get_energy(handle, &readings, 1000);
        if (err == ESP_OK)
        {
            ESP_LOGI(TAG, "Got readings: T1 = %lu, T2 = %lu", readings.t1, readings.t2);
        }
        else
        {
            ESP_LOGE(TAG, "Error getting readings: 0x%x", err);
        }

        ESP_LOGI(TAG, "Free heap %lu ", esp_get_free_heap_size());

        vTaskDelay(5000 / portTICK_PERIOD_MS);
    }
}


ce2727a.c
#include <stdio.h>
#include <string.h>
#include "ce2727a.h"
#include "frames.h"
#include "esp_log.h"
#include "esp_check.h"
#include "driver/i2c.h"
#include "driver/uart.h"

#include "frames.h"

uint16_t crc16iec(const uint8_t *buffer, uint16_t len);

static const char *TAG = "ce2727a-driver";

#define CHECK(x)                \
    do                          \
    {                           \
        esp_err_t __;           \
        if ((__ = x) != ESP_OK) \
            return __;          \
    } while (0)

#define MAGIC_2 0x02

#define COM_ENQ 0x01
#define COM_REC 0x03
#define COM_DRJ 0x0A
#define COM_OK 0x0B

#define ENQ_CMD_INFO 0x00
#define ENQ_CMD_DATE_TIME 0x01
#define ENQ_CMD_ACTIVE_POWER 0x02
#define ENQ_CMD_CONSUMED_ENERGY 0x03

// driver state
struct ce2727a_driver_state_t
{
    ce2727a_config_t config;
    ce2727a_request_command_t *tx_buffer;
    uint16_t tx_buffer_size;
};

uart_config_t default_uart_config = {
    .baud_rate = 9600,
    .data_bits = UART_DATA_8_BITS,
    .parity = UART_PARITY_EVEN,
    .stop_bits = UART_STOP_BITS_1,
    .flow_ctrl = UART_HW_FLOWCTRL_DISABLE};

esp_err_t ce2727a_init(const ce2727a_config_t *config, ce2727a_handle_t *out_handle)
{
    struct ce2727a_driver_state_t *driver_state = calloc(1, sizeof(struct ce2727a_driver_state_t));
    if (driver_state == NULL)
    {
        return ESP_ERR_NO_MEM;
    }
    driver_state->tx_buffer = calloc(1, sizeof(ce2727a_request_command_t));
    if (driver_state->tx_buffer == NULL)
    {
        return ESP_ERR_NO_MEM;
    }
    driver_state->tx_buffer_size = sizeof(ce2727a_request_command_t);
    driver_state->config = *config;
    *out_handle = driver_state;

    ESP_ERROR_CHECK(uart_param_config(config->uart_port, (config->uart_config_override != NULL ? config->uart_config_override : &default_uart_config)));
    ESP_ERROR_CHECK(uart_set_pin(config->uart_port, config->uart_tx_pin, config->uart_rx_pin, config->uart_de_pin, UART_PIN_NO_CHANGE));
    ESP_ERROR_CHECK(uart_driver_install(config->uart_port, 256, 0, 0, NULL, 0));
    ESP_ERROR_CHECK(uart_set_mode(UART_NUM_2, UART_MODE_RS485_HALF_DUPLEX));

    return ESP_OK;
}

esp_err_t ce2727a_deinit(ce2727a_handle_t handle)
{
    uart_driver_delete(handle->config.uart_port);
    free(handle->tx_buffer);
    free(handle);
    return ESP_OK;
}

void ce2727a_fill_enq_cmd(ce2727a_handle_t handle, uint8_t com_id)
{
    ce2727a_request_command_t *f = handle->tx_buffer;
    f->header.two = MAGIC_2;
    f->header.length = sizeof(ce2727a_request_command_t);
    f->header.address = handle->config.addr;
    f->header.password = handle->config.password;
    f->header.com = COM_ENQ;
    f->header.id = (uint8_t)com_id;
    f->crc16 = crc16iec((const uint8_t *)f, sizeof(ce2727a_request_command_t) - 2); // minus 2 bytes for CRC
}

esp_err_t ce2727a_send_enq_cmd(ce2727a_handle_t handle, uint8_t com_id, uint16_t timeout_ms)
{
    ce2727a_fill_enq_cmd(handle, com_id);
    // Check initialized
#if HELPER_TARGET_IS_ESP32 && ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 0, 0)
    if (!uart_is_driver_installed(handle->config.uart_port))
        return ESP_ERR_INVALID_STATE;
#endif

    ESP_ERROR_CHECK(uart_flush(handle->config.uart_port));
    int bytes_pushed = uart_write_bytes(handle->config.uart_port, (char *)handle->tx_buffer, handle->tx_buffer_size);
    if (bytes_pushed != handle->tx_buffer_size)
    {
        ESP_LOGE(TAG, "uart_write_bytes couldn't write frame. shall be: %d, written: %d", handle->tx_buffer_size, bytes_pushed);
        return ESP_ERR_INVALID_STATE;
    }
    return ESP_OK;
}

esp_err_t ce2727a_receive_proper_response(ce2727a_handle_t handle, uint16_t expected_response_size, uint16_t timeout_ms)
{
    memset(handle->config.rx_buffer, 0, handle->config.rx_buffer_size);

    // Read response from serial buffer
    int len = uart_read_bytes(handle->config.uart_port, handle->config.rx_buffer,
                              handle->config.rx_buffer_size,
                              timeout_ms / portTICK_PERIOD_MS);

    if (len < expected_response_size)
    {
        return ESP_ERR_TIMEOUT;
    }

    if (len != expected_response_size)
    {
        return ESP_ERR_INVALID_RESPONSE;
    }

    u_int16_t *p_crc = (u_int16_t *)(handle->config.rx_buffer + expected_response_size - 2);
    if (crc16iec(handle->config.rx_buffer, expected_response_size - 2) != *p_crc)
    {
        return ESP_ERR_INVALID_CRC;
    }

    return ESP_OK;
}

esp_err_t ce2727a_get_active_power(ce2727a_handle_t handle, uint32_t *active_power, uint16_t timeout_ms)
{
    if (handle->config.rx_buffer_size < sizeof(ce2727a_response_active_power_t) || active_power == NULL)
    {
        return ESP_ERR_INVALID_ARG;
    }
    esp_err_t err = ce2727a_send_enq_cmd(handle, ENQ_CMD_ACTIVE_POWER, timeout_ms);
    if (err != ESP_OK)
        return err;
    err = ce2727a_receive_proper_response(handle, sizeof(ce2727a_response_active_power_t), timeout_ms);
    ce2727a_response_active_power_t *res = (ce2727a_response_active_power_t *)handle->config.rx_buffer;

    *active_power = res->activePower;

    return err;
}

esp_err_t ce2727a_get_energy(ce2727a_handle_t handle, ce2727a_readings_energy_t *readings, uint16_t timeout_ms)
{
    if (handle->config.rx_buffer_size < sizeof(ce2727a_response_consumed_energy_t) || readings == NULL)
    {
        return ESP_ERR_INVALID_ARG;
    }
    esp_err_t err = ce2727a_send_enq_cmd(handle, ENQ_CMD_CONSUMED_ENERGY, timeout_ms);
    if (err != ESP_OK)
        return err;
    err = ce2727a_receive_proper_response(handle, sizeof(ce2727a_response_consumed_energy_t), timeout_ms);
    ce2727a_response_consumed_energy_t *res = (ce2727a_response_consumed_energy_t *)handle->config.rx_buffer;

    readings->active_tariff = res->currentTariff;
    readings->total = res->total;
    readings->t1 = res->t1;
    readings->t2 = res->t2;
    readings->t3 = res->t3;
    readings->t4 = res->t4;

    return err;
}

//---------------------------------------------------------------------------------------
//  The standard 16-bit CRC polynomial specified in ISO/IEC 3309 is used.
//             16   12   5
//  Which is: x  + x  + x + 1
//----------------------------------------------------------------------------
uint16_t crc16iec(const uint8_t *buffer, uint16_t len)
{
    uint16_t crc = 0xffff;
    uint8_t d;
    do
    {
        d = *buffer++ ^ (crc & 0xFF);
        d ^= d << 4;
        crc = (d << 3) ^ (d << 8) ^ (crc >> 8) ^ (d >> 4);
    } while (--len);
    crc ^= 0xFFFF;
    return crc;
}
